specs/codecs/wac/spec.md

```diff
@@ -0,0 +1,192 @@
+
+Every type value can be parsed knowing only the type and without any outside context
+like the container or positional delimiters.
+
+#### Tokens
+
+TODO: Make constants in code match spec
+
+Note: The existing implementations of WAC in https://github.com/aschmahmann/wasm-ipld (Go + Rust/WASM) have different token numbers that we'll need to synchronize before merge.
+
+The tokens for the codec roughly correspond with the [IPLD Data Model Kinds]
+
+They are:
+	Null   = 0
+	True   = 1
+	False  = 2
+	Int    = 3
+	NInt   = 4
+	Float  = 5
+	String = 6
+	Bytes  = 7
+	List   = 8
+	Map    = 9
+	Link   = 10
+
+| Int | Token |
+|---|---|
+| 0 | TYPE_LINK |
+| 1 | TYPE_INTEGER |
+| 2 | TYPE_NEGATIVE_INTEGER |
+| 3 | TYPE_FLOAT |
+| 4 | TYPE_NEGATIVE_FLOAT |
+| 5 | TYPE_STRING |
+| 6 | TYPE_BINARY |
+| 7 | TYPE_MAP |
+| 8 | TYPE_LIST |
+| 9 | VALUE_NULL |
+| 10 | VALUE_TRUE |
+| 11 | VALUE_FALSE |
+
+#### TYPE_LINK
+
+```
+| 0 | CID |
+```
+
+Note: Simple-DAG put a VARINT_LENGTH before the CID indicating how long it was.
+
+CIDs are self-delimiting so this didn't seem necessary.
+
+#### TYPE_INTEGER
+
+```
+| 1 | VARINT |
+```
+
+#### TYPE_SIGNED_INTEGER
+
+```
+| 2 | VARINT |
+```
+
+#### TYPE_FLOAT
+
+```
+| 3 | MATISSA_LENGTH | VARINT
+```
+
+TODO: Floats (and negative floats) need definition here.
+Making implementations actually bidirectional with respect to the IPLD Data Model seems difficult here.
+
+#### TYPE_NEGATIVE_FLOAT
+
+```
+| 4 | MATISSA_LENGTH | VARINT
+```
+
+#### TYPE_STRING
+
+```
+| 5 | VARINT_LENGTH | STRING
+```
+
+Note: This is essentially the same as Binary, but with a different token flag.
+
+As with the IPLD Data Model itself Strings are basically just Bytes with something to tag them as "string-like" but where UTF-8 is encouraged. See [Strings].
+
+#### TYPE_BINARY
+
+```
+| 6 | VARINT_LENGTH | BINARY
+```
+
+#### TYPE_MAP
+
+```
+| 7 | VARINT_NUM_PAIRS | PAIRS
+```
+
+`PAIRS` contains alternating keys then values concatenated. i.e. ` KEY1 | VALUE1 | KEY2 | VALUE2 ...`
+
+This codec does not have any form of canonical map sorting as that would make it ill-fitted to the IPLD Data Model.
+
+As in the IPLD Data Model map keys must be of type String, however as described in the String section the only distinction between Strings and Bytes are identifier hints.
+
+Note: Simple-DAG went with `KEYS_VARINT_LENGTH | VALUES_VARINT_LENGTH | KEYS | VALUES |`. Both seem doable,
+this approach seemed to make writing encoder/decoders really simple. However, adding in more length prefixes
+makes creating faster "zero copy" decoders very nice. It seems to mostly be a tradeoff for which side has to have bigger buffers, the encoder or the decoder.
+
+Note: We could assert that keys are just `VARINT_LENGTH | STRING` and remove the String token since it's always a string.
+It's some added complexity and really stops us from putting anything other than Strings in map keys, but that may not be too bad.
```

Comment:
My understanding of the Data Model is we're restricted to "Strings" as map keys even if they're not UTF-8 strings. Keeping the extra token here seems safe, but could be convinced otherwise. It seems like keeping the token allows us room to maneuver in the future if we extend what are valid map keys, although to be fair we could just make a new codec version at that time 🤷.

Reply:
Either way you should specify - I assumed from first read that this _was_ strictly a string so the extra token was unnecessary - I think SimpleDAG did this?
I'd prefer leaving off the token and making the assumption. I _think_ strictly string keys at the data model layer is going to stay that way, but other key types above that might become appropriate. So for a codec, it seems safe to me to lock in strings.